runtime/lua/vim/_meta.lua

```diff
@@ -0,0 +1,455 @@
+    end
+
+    seen[v] = true
+  end
+
+  return result
+end
+
+local remove_value = function(t, val)
+  if vim.tbl_islist(t) then
+    local remove_index = nil
+    for i, v in ipairs(t) do
+      if v == val then
+        remove_index = i
+      end
+    end
+
+    if remove_index then
+      table.remove(t, remove_index)
+    end
+  else
+    t[val] = nil
+  end
+
+  return t
+end
+
+local value_mutator = function(current, new, mutator)
+  if singular_values[type(current)] then
+    error(
+      "This is not possible to do. Please do something different: "
+      .. tostring(current)
+      .. " // "
+      .. tostring(new)
+    )
+  end
+
+  if type(new) == 'string' then
+    if vim.tbl_islist(current) then
+      mutator.list(current, new)
+    else
+      set_key_value(current, new)
+    end
+
+    return current
+  elseif type(new) == 'table' then
+    if vim.tbl_islist(current) then
+      assert(vim.tbl_islist(new))
+      vim.list_extend(current, new)
+    else
+      assert(not vim.tbl_islist(new), vim.inspect(new) .. vim.inspect(current))
+      current = vim.tbl_extend("force", current, new)
+    end
+
+    return current
+  else
+    error("Unknown type")
+  end
+end
+
+local prepend_value = function(current, new)
+  return value_mutator(current, new, {
+    list = function(left, right)
+      table.insert(left, 1, right)
+    end,
+  })
+end
+
+local add_value = function(current, new)
+  return value_mutator(current, new, {
+    list = function(left, right)
+      table.insert(left, right)
+    end,
+  })
+end
+
+local convert_lua_to_vimoption = function(t)
+  if type(t) ~= 'table' then
+    return t
+  elseif vim.tbl_islist(t) then
+    t = remove_duplicate_values(t)
+
+    return table.concat(t, ',')
+  else
+    return concat_key_values(t, ',', ':')
+  end
+end
+
+local clean_value = function(v)
+  if singular_values[type(v)] then
+    return v
+  end
+
+  local result = v:gsub('^,', ''):gsub(',$', '')
+
+  return result
+end
+
+local literal_keys = {
+  _value = true,
+  _option = true,
+}
+
+local meta_keys = {
+  prepend = true,
+  append = true,
+  remove = true,
+  set = true,
+}
+
+local opt_mt
+opt_mt = {
+  __index = function(t, k)
+    if literal_keys[k] then
+      return rawget(t, k)
+    end
+
+    if meta_keys[k] then
+      return rawget(opt_mt, k)
+    end
+
+    return setmetatable({ _option = k, }, opt_mt)
+  end,
+
+  set = function(self)
+    local k = assert(self._option, "Must have an option")
+    local raw_v = assert(self._value, "Must have a value to set")
+    local v = convert_lua_to_vimoption(raw_v)
+
+    -- TODO: Determine how much of this we should do here
+    -- local option_info = assert(all_options_info[k], "Must be a valid option name")
+    -- if type(v) ~= option_info.type then
+    --   error(string.format("Invliad type. Expected Type: %s, Got: `%s`", option_info.type, tostring(v)))
+    -- end
+
+    -- TODO: Figure out why nvim_set_option doesn't override values the same way.
```

Comment:
`nvim_set_option` should be used if the option has "global" scope _or_ buf/win scope with "global_local" flag set. otherwise use nvim_buf/win_set_option accordingly. at me again with whatever cases are still broken :]